### PR TITLE
C++: IR dataflow through varargs

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR fixes https://github.com/github/codeql-c-analysis-team/issues/35 by implementing dataflow through vararg functions.

For the following function:

```
void use_varargs(int ignored, ...)
{
  __builtin_va_list ap;
  __builtin_va_start(ap, ignored);
  const char* buf = __builtin_va_arg(ap, const char*);
  int i = __builtin_va_arg(ap, int);
  __builtin_va_end(ap);
}
```
we create the following IR (+ stuff I removed):

```
#   31| char* use_varargs(int)
#   31|   Block 0
...
#   33|     r33_1(glval<__va_list_tag[1]>) = VariableAddress[ap]          :
#   33|     m33_2(__va_list_tag[1])        = Uninitialized[ap]            : &:r33_1
#   34|     r34_1(glval<__va_list_tag[1]>) = VariableAddress[ap]          :
#   34|     r34_2(__va_list_tag *)         = Convert                      : r34_1
#   34|     r34_3(glval<int>)              = VariableAddress[ignored]     :
#   34|     v34_4(void)                    = VarArgsStart                 : 0:r34_2, 1:r34_3
#   35|     r35_1(glval<char *>)           = VariableAddress[buf]         :
#   35|     r35_2(glval<__va_list_tag[1]>) = VariableAddress[ap]          :
#   35|     r35_3(__va_list_tag *)         = Convert                      : r35_2
#   35|     r35_4(glval<char *>)           = VarArg                       : 0:r35_3
#   35|     r35_5(char *)                  = Load                         : &:r35_4, ~m31_4
#   35|     m35_6(char *)                  = Store                        : &:r35_1, r35_5
#   36|     r36_1(glval<int>)              = VariableAddress[i]           :
#   36|     r36_2(glval<__va_list_tag[1]>) = VariableAddress[ap]          :
#   36|     r36_3(__va_list_tag *)         = Convert                      : r36_2
#   36|     r36_4(glval<int>)              = VarArg                       : 0:r36_3
#   36|     r36_5(int)                     = Load                         : &:r36_4, ~m31_4
...
```
and given a call: `use_varargs(0, "test", 42)` data will flow from `"test"` and `42` to the "initializer"
```
#   34|     v34_4(void)                    = VarArgsStart                 : 0:r34_2, 1:r34_3
```
and from that instruction to `r35_4` and `r36_4`.

I don't really like how the arguments flow to an instruction with a `void` type, but I couldn't think of a better common place for them to flow (to avoid `number of arguments * number of uses` data flow edges).

Currently this PR also ignores the reads and writes to the `__builtin_va_list` argument as the IR doesn't generate memory operands for vararg instructions.

### CODE REVIEW

**Comment by jbj** [comment]
> I don't really like how the arguments flow to an instruction with a `void` type, but I couldn't think of a better common place for them to flow (to avoid `number of arguments * number of uses` data flow edges).

I'd like to see the codebase where this causes a real performance problem (or perhaps I wouldn't like to see it). I assume it's fine for a function to have multiple `va_start` calls (in different branches of an `if`), so in principle we could still get a quadratic number of edges here. But in practice it seems unlikely to be a problem.

My initial feeling is therefore that the data-flow edges should go to the `va_arg` calls.

**Comment by MathiasVP** [comment]
> I'd like to see the codebase where this causes a real performance problem (or perhaps I wouldn't like to see it). I assume it's fine for a function to have multiple `va_start` calls (in different branches of an `if`), so in principle we could still get a quadratic number of edges here. But in practice it seems unlikely to be a problem.

That's a good point. My fear is that someone might have something like:

```
switch(...) {
  case 1: ... = va_arg(pa, int*); break;
  case 2: ... = va_arg(pa, char*); break;
  ... // a thousand different cases with different types here ...
}
```

but I haven't found anything bigger than this so far: https://github.com/mpaland/printf/blob/master/printf.c (which has ~20 occurrences of `va_arg` in one function).

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 1]
```diff
@@ -166,6 +170,16 @@ private class ThisParameterNode extends InstructionNode {
   override string toString() { result = "this" }
 }
 
+class VarArgParameterNode extends ParameterNode {
+  override VarArgsStartInstruction instr;
+
+  override predicate isParameterOf(Function f, int i) {
+    i instanceof ParameterIndex and
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Can we restrict `i` even more? Like restricting it to the max argument number passed to this particular function? (can virtually-dispatched functions use varargs?)

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 2]
```diff
@@ -323,7 +337,7 @@ predicate simpleLocalFlowStep(Node nodeFrom, Node nodeTo) {
   simpleInstructionLocalFlowStep(nodeFrom.asInstruction(), nodeTo.asInstruction())
 }
 
-private predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction iTo) {
+predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction iTo) {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Why no longer private?

**Comment by jbj** [comment]
How are we supporting `vprintf`-like functions? I'd assume that most `printf` implementations are wrappers for `vprintf`, so that should create a natural "funnel" to prevent the "arguments * uses" cross product.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Oops! That was a leftover from my debugging that slipped into a commit.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 3]
```diff
@@ -142,18 +142,22 @@ class ExprNode extends InstructionNode {
   override string toString() { result = this.asConvertedExpr().toString() }
 }
 
+abstract class ParameterNode extends InstructionNode {
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Gotcha. I will convert this to a non-abstract class.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Looks like it: https://ideone.com/n2qdFZ

**Comment by MathiasVP** [comment]
> How are we supporting `vprintf`-like functions? I'd assume that most `printf` implementations are wrappers for `vprintf`, so that should create a natural "funnel" to prevent the "arguments * uses" cross product.

That's a good point. To support flow from `getenv("HOME")` to `buf` in:

```c
char* foo(int ignored, ...)
{
  char* buf = (char*)malloc(16);
  __builtin_va_list ap;
  __builtin_va_start(ap, ignored);
  vsprintf(buf, "%s", ap);
  __builtin_va_end(ap);
  return buf;
}
void bar() {
  char* path = foo(0, getenv("HOME"));
}
```

we need flow from `getenv("HOME")` to <i>something</i>, followed by `__builtin_va_start` transferring flow that <i>something</i> to the memory pointed to by `ap`.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll (hunk 1 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Okay. Then I'm fine with keeping `i instanceof ParameterIndex` for now.

File: cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: The abstract predicate also needs to become non-abstract. Example:

```
predicate isParameterOf(Function f, int i) { none() } // overridden in subclasses
```

**Comment by jbj** [comment]
Okay, so it looks like the idea of sending flow from arguments to `va_arg` is out.

Taking a step back, what would be the ideal IR code generation for the `__builtin_va_*` calls?
1. Ideally we want to track not only values but also indirections. Just like each parameter is initialized by two instructions, one for the value and one for the indirection, perhaps we also need the varargs to be initialized by two instructions.
2. Ideally, the flow should be evident from the memory edges and instruction superclasses without further modelling in the data-flow library. That means it'll be visible not just to the data-flow library but to any other consumer of the IR, including `aliased_ssa`. If `va_list` can be modelled as a stack-allocated array, and `va_arg` can be modelled as indexing into that array, then it seems to me like `va_arg` could be a `CopyInstruction` that does a non-exact load from the address of the given `va_list`. Except I think that would be unsound for `ValueNumbering` since two consecutive calls to `va_arg` can return different things.
3. The whole mapping of arguments to parameters and return values to outputs is getting complicated. In the near future, I'd like to pull it out as a separate library that can be shared with [the alias analysis](https://github.com/Semmle/ql/blob/3ae1aada372b668b5cbd6314995da6787f4b5fed/cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasAnalysis.qll#L199-L214) and other interprocedural libraries.

All of those thoughts don't necessarily fit into this PR. I'll follow up with something more pragmatic.

**Comment by MathiasVP** [comment]
> Okay, so it looks like the idea of sending flow from arguments to `va_arg` is out.
> 
> Taking a step back, what would be the ideal IR code generation for the `__builtin_va_*` calls?

I'm currently trying the following idea:

- Each vararg function gets a new instruction, `InitializeVarArgsInstruction`, that receives dataflow from all of the incoming varargs, and returns a new memory.
- A `VarArgStartInstruction` consumes that memory and outputs a new memory containing the memory pointed to by the `__builtin_va_list` argument.
- Finally, each `va_arg` expression introduces a `VarArgInstruction` along with a read and write side effect.

But I'm open to alternatives. If nothing else, this is a nice way for me to learn more about how the IR is constructed.

**Comment by jbj** [comment]
Sounds good!

**Comment by dbartol** [comment]
What's the flow that's most important to capture? Is it the flow the values of an individual argument to the point where it's loaded via `va_arg()`? Or is it the flow from the whole bundle of arguments to the `...` through a few levels of calls that take `va_list` arguments directly? The latter is easier than the former, because you can just track the argument bundle as if it were a stack-allocated struct of unknown size, along the lines of what @jbj suggested above.

What if we pretended that a call to a vararg function was desugared from this:
```cpp
printf("%d %f %s", 1, 3.14, "foo");
```
to this:
```cpp
struct Temp { int a; double b; const char* c; } __args = { 1, 3.14, "foo" };
printf("%d %f %s", &__args);
```
On the callee side, it would be desugared from:
```cpp
void printf(const char* fmt, ...) {
  va_list args;
  va_start(args, fmt);
  while (???) {
    switch (???) {
    case ???:
      int d = va_arg(args, int);
      ???
    case ???:
      double f = va_arg(args, double);
      ???
    case ???:
      const char* s = va_arg(args, const char*);
      ???
    }
  }
  va_end(args);
}
```
to
```cpp
void printf(const char* fmt, const unsigned char* __ellipsis) {
  const unsigned char* args;
  args = __ellipsis;  // va_start
  while (???) {
    switch (???) {
    case ???:
      int d = *reinterpret_cast<const int*>(args);
      args += sizeof(int);
      ???
    case ???:
      double f = *reinterpret_cast<const double*>(args);
      args += sizeof(double);
      ???
    case ???:
      const char* s = *reinterpret_cast<const char* const*>(args);
      args += sizeof(const char*);
      ???
    }
  }
}
```
If you ignore alignment and some details about stack frame setup, this is actually pretty close to what real codegen for varargs looks like.

> Each vararg function gets a new instruction, InitializeVarArgsInstruction, that receives dataflow from all of the incoming varargs, and returns a new memory.

How about an `InitializeVarArgs` instruction that returns a pointer to the beginning of the ellipsis argument buffer (`&__args` in the desugared call above), followed by an `InitializeIndirectVarArgs` that initializes the ellipsis argument buffer itself (`__args` in the desugared call above). These two instructions work almost exactly like `InitializeParameter` and `InitializeIndirectParameter`.

> A VarArgStartInstruction consumes that memory as a total load and outputs a new memory containing the memory pointed to by the __builtin_va_list argument.

I don't think that `VarArgStart` really need to consume the memory defined by `InitializeVarArgs`. In fact, I think this can just be a `Store` of the result of `InitializeVarArgs` into the `va_list` variable.

> Finally, each va_arg expression introduces a VarArgInstruction along with a read and write side effect.

I think this can just be a `Load` from the ellipsis buffer, followed by some sort of addition to move the `va_list` to point to the next argument.

For wrapper functions that just pass their varargs through via `va_list`, this amounts transforming this:
```cpp
void printf(const char* fmt, ...) {
  va_list args;
  va_start(args, fmt);
  vprintf(fmt, args);
}
```
to:
```cpp
void printf(const char* fmt, const unsigned char* __ellipsis) {
  const unsigned char* args = __ellipsis;
  vprintf(fmt, args);
}
```
which looks exactly like any other wrapper function. We'd get flow from `*__ellipsis` into `vprintf()` more or less automatically.

**Comment by jbj** [comment]
I'm wary about desugaring so much that we're changing the number and meaning of arguments and parameters. The abstraction of the IR is not so complete that it replaces the AST, and queries still have to flip back and forth between the IR and the AST to get a complete picture of `Declaration`s.

**Comment by dbartol** [comment]
Reading my comment again, I realize that I didn't make it clear that I don't think we should _actually_ desugar on the caller side. Passing the ellipsis arguments as argument operands to the `Call` instruction seems totally reasonable. It's on the callee side where I think we should desugar; that's the side where we don't have any idea how many actual arguments we had anyway.

**Pull request closed without merging**